### ISSUE TITLE
Adding peer discovery using Kad DHT

### ISSUE THREAD

**Issue by robzajac** [libp2p/py-libp2p#128]
py-libp2p should implement peer discovery based on the interfaces added in #123 and conforming to standard libp2p implementations of peer discovery (generally using the Kademlia DHT, see [go-libp2p](https://github.com/libp2p/go-libp2p-kad-dht) and [js-libp2p](https://github.com/libp2p/js-libp2p-kad-dht)).

@zaibon we saw that you had some work around this in #123. Are you still planning on submitting a PR, and if so on what timeline? We're happy to help and would welcome the addition. Thanks!

**Comment by zaibon** [comment]
Yep I had some code that I wrote when I started looking into the kadmelia module.
It actually doesn't implement exactly the same interface but with a bit of work I can make it comply to the interfaces you defined.

I'll open a draft PR with my current code, so we can discuss further.

**Comment by amirouche** [comment]
I started a new dht implementation based on @bmuller rpcudp and kademlia. I reworked the whole kademlia module based on my understanding of the kademlia publication. In particular, I changed the code that does value lookups. [In the kademlia project, the algorithm is not apparent at all](https://github.com/bmuller/kademlia/blob/master/kademlia/crawling.py#L71), here is my take on it:

https://github.com/amirouche/qadom/blob/c4a5f05e1c2138e486b1d27640d941d50f30c745/qadom/peer.py#L301-L339

It somewhat long, I acknowledge it. Factorizing leads to less readable code. Another important thing I changed vs. the kademlia publication is that it is the **sole** responsibility of the publisher to refresh values (aka. call `STORE`). That part is not in my codebase as of yet for some dubious reason.

Hope this helps. Good luck.
